docs/design/settings/extensions/modular-settings-extensions.md

```diff
@@ -0,0 +1,314 @@
+`schnauzer` fetches the entirety of the settings tree, defines some useful helper functions, and then uses these as context for rendering the template.
+With the introduction of settings extensions, this will change such that templates specify `(name, version)` tuples for all extensions needed to render the template.
+These extension specifications will be given within the template as "front matter" -- a TOML block specified at the head of the document and delineated with `---` as such:
+
+```toml
+---
+[required-extensions]
+hostname = v1
+kubernetes = v1
+---
+# Below here the template is written as usual
+```
+
+This will allow Bottlerocket to find the owning extensions for the data needed to render the template.
+It also tightly binds the template to a given implementation of the requisite settings, making it much more difficult to accidentally break "downstream" templates with "usptream" changes.
+
+`schnauzer` can also continue to be invoked as a setting generator without the need for a templated configuration file.
+In those circumstances, you are still required to list your extension dependencies and their version, but these can be done as command line arguments:
+
+```bash
+$ schnauzer --required-extension 'hostname=v1' '{{ network.hostname }}'
+```
+
+As an optimization, templates which render to `systemd` units will invoke a restart command to systemd
+
+#### Extending Handlebars Helpers
+Handlebars has a few utilities for applying simple processes to data in order to format it appropriately for your document; however, you often need to implement your own "helpers" for more complex text formatting.
+This is currently done in Bottlerocket by adding helper implementations to the `schnauzer` tool directly.
+Bottlerocket settings extensions will expose versioned handlebar helpers:
+
+```bash
+# hostname.toml setting extension config file
+[templating.v1] # template helpers are versioned
+helpers = [
+    "localhost_aliases"
+]
+
+
+$ ./hostname proto1 template-helper --version v1 --helper localhost_aliases 'arg1' 'arg2'
+#  rendered output returned as a string on stdout
+```
+
+Within the handlerbars template, the helper name will be prepended with `$EXTENSION_NAME.` to avoid collisions with other settings extensions.
+As an example, to use the `localhost_aliases` helper above from a template, your template may look something like this:
+
+```
+{{ hostname.localhost_aliases(arg1, arg2) }}
+```
+
+### Settings Generation
+Bottlerocket settings currently have the ability to compute a default value via something called a "settings generator".
+During boot, settings extensions are given a chance to populate settings with default values that are then persisted, effectively determining the setting value for the lifetime of that instance.
+
+Settings generation often depends on the presence of other settings; therefore the invocation protocol will allow settings generation to accept rendered templates as input:
```

Comment:
rendered templates instead of other settings? wouldn't be more convenient for me to say I need `aws.region` than to find a template which contains that value?

Reply:
This design made `aws.region` more opaque from the apiserver's perspective, so it wouldn't be possible for the server to resolve that. It's one of the reasons why we'd need to move system services to use template files. Basically anything that would request a setting would do so via a template.